library/src/main/java/com/wultra/android/mtokensdk/operation/OperationsService.kt

```diff
@@ -163,6 +170,34 @@ class OperationsService: IOperationsService {
         }
     }
 
+    private fun processServerTime(response: OperationListResponse, requestStarted: ZonedDateTime) {
+
+        val now = ZonedDateTime.now()
+        val requestDelayMilliseconds = now.toInstant().toEpochMilli() - requestStarted.toInstant().toEpochMilli()
+
+        // server does not support this feature
+        if (response.currentTimestamp == null) {
+            return
+        }
+
+        // Reject the value if the request took too long and we already have a server date.
+        // This is to avoid volatility of the value
+        if (serverDateShiftInMilliSeconds != null && requestDelayMilliseconds > 1_000) {
+            return
+        }
+
+        // We're adding half of the time that the request took to compensate for the network delay
+        val serverTime = response.currentTimestamp.plus((requestDelayMilliseconds/2), ChronoUnit.MILLIS)
+
+        // If the difference is under 0.3 seconds, we ignore the new value to avoid unnecessary changes that might be due to network delay.
+        val currentServerDate = currentServerDate()
+        if (currentServerDate != null && abs((currentServerDate.toInstant().toEpochMilli() - serverTime.toInstant().toEpochMilli())) > 300) {
+            return
+        }
```

Comment:
I think this is wrong. If we have already adjusted date (e.g. currentServerDate is not null) and I change the time on the device manually, then this condition don't catch that change, because it rejects difference greater than 300 ms.

I think the correct algorithm should be like this:

1. measure the time of request - response roundtrip - `requestDelayMilliseconds` ✅
2. calculate `serverTime` as received server time value plus `requestDelayMilliseconds` divided by 2 ✅
3. calculate difference between local time and the calculated `serverTime`
4. If the absolute difference is greater than some threshold (like 10 second, 1 minute, etc..) then we suppose to keep this in `serverDateShiftInMilliSeconds`
5. Optionally, to prevent the time variation caused by multiple requests, we should add an additional threshold that compare old and new `serverDateShiftInMilliSeconds`. If the difference is less than a threshold (for example 1 second), then we can ignore this.

The value in step 4 depends on the purpose of the adjusted time. For example, in PowerAuth protocol, if the time window set on the server to accept the calculated token digest is set to 1 minute, then we're OK to set threshold to 30 seconds. But in general, I think than 10 seconds should be OK for any purpose.

Reply:
I checked the iOS implementation and the condition there is OK. But I think that it's not a good idea to take already adjusted time into the calculation whether we should adjust the time again. It should always compare the real time on the device against the time on the server.

Reply:
I hope the last note :) The filter that reject server time from the request that took too long to complete is OK.

Reply:
OK, I finally got back to this and indeed there is a wrong `>` which I fixed.

I modified the rest (slightly different than you described) to take the device time change into account better.